File: easy-file-choose/src/main/java/com/classic/file/choose/EasyFileChoose.java
package com.classic.file.choose;

import android.app.Activity;
import android.content.Intent;
import android.support.annotation.NonNull;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.lang.ref.WeakReference;
import java.util.Comparator;

/**
 * 文件选择
 *
 * @author Classic
 * @version v1.0, 2017/2/26 4:02 PM
 */
public class EasyFileChoose {
    static final String INTENT_KEY_PATH = "path";
    private String mTitle;
    private String mDefaultPath;
    private WeakReference<Activity> mActivity;
    private static FilenameFilter sNameFilter;
    private static FileFilter sFileFilter = new DefaultFileFilter();
    private static Comparator<File> sFileComparator = new DefaultFileComparator();

    public static String getPath(Intent intent) {
        return null == intent || !intent.hasExtra(INTENT_KEY_PATH) ?
            "" : intent.getStringExtra(INTENT_KEY_PATH);
    }

    static FilenameFilter nameFilter() {
        return sNameFilter;
    }

    static FileFilter fileFilter() {
        return sFileFilter;
    }

    static Comparator<File> fileComparator() {
        return sFileComparator;
    }

    private EasyFileChoose(Builder builder) {
        mTitle = builder.title;
        mDefaultPath = builder.path;
        mActivity = new WeakReference<>(builder.activity);
        if (null != builder.nameFilter) sNameFilter = builder.nameFilter;
        if (null != builder.fileFilter) sFileFilter = builder.fileFilter;
        if (null != builder.fileComparator) sFileComparator = builder.fileComparator;
    }

    @SuppressWarnings("unused")
    public void choose(int requestCode) {
        Activity activity = mActivity.get();
        if (activity != null) {
            FileChooseActivity.start(activity, mTitle, mDefaultPath, requestCode);
        }
    }

    @SuppressWarnings("unused")
    public static final class Builder {
        private String title;
        private String path;
        private Activity activity;
        private FilenameFilter nameFilter;
        private FileFilter fileFilter;
        private Comparator<File> fileComparator;

        public Builder() {}

        public Builder title(@NonNull String val) {
            title = val;
            return this;
        }

        public Builder path(@NonNull String val) {
            path = val;
            return this;
        }

        public Builder activity(@NonNull Activity val) {
            activity = val;
            return this;
        }

        public Builder filenameFilter(@NonNull FilenameFilter val) {
            nameFilter = val;
            fileFilter = null;
            return this;
        }

        public Builder fileFilter(@NonNull FileFilter val) {
            fileFilter = val;
            nameFilter = null;
            return this;
        }

        public Builder fileComparator(@NonNull Comparator<File> val) {
            fileComparator = val;
            return this;
        }

        public EasyFileChoose build() {return new EasyFileChoose(this);}
    }
}

File: easy-file-choose/src/main/java/com/classic/file/choose/EasyFileChoose.kt
package com.classic.file.choose

import android.app.Activity
import android.content.Intent
import android.os.Environment
import java.io.File
import java.io.FileFilter
import java.io.FilenameFilter
import java.util.Comparator

/**
 * 文件选择
 *
 * @author Classic
 * @version v1.0, 2017/2/26 4:02 PM
 */
@Suppress("unused")
object EasyFileChoose {
    private var directoryIcon: Int = R.drawable.ic_easy_file_choose_directory
    private var fileIcon: Int = R.drawable.ic_easy_file_choose_file
    private var selectedColor: Int = R.color.EasyFileChooseSelected
    private var transparentColor: Int = R.color.EasyFileChooseTransparent
    internal const val INTENT_KEY_PATH = "path"
    internal var defaultPath: String = Environment.getExternalStorageDirectory().absolutePath
    internal var title: Int = R.string.easy_file_choose_title
    internal var nameFilter: FilenameFilter? = null
    internal var fileFilter: FileFilter? = null
    internal var comparator: Comparator<File> = DefaultFileComparator()

    /**
     * 设置起始路径
     */
    fun setPath(value: String): EasyFileChoose {
        defaultPath = value
        return this
    }

    /**
     * 设置标题
     */
    fun setTitle(resId: Int): EasyFileChoose {
        title = resId
        return this
    }

    /**
     * 设置文件图标
     */
    fun setFileIcon(resId: Int): EasyFileChoose {
        fileIcon = resId
        return this
    }

    /**
     * 设置目录图标
     */
    fun setDirectoryIcon(resId: Int): EasyFileChoose {
        directoryIcon = resId
        return this
    }

    /**
     * 设置文件选中的背景颜色
     */
    fun setSelectedBackgroundColor(resId: Int): EasyFileChoose {
        selectedColor = resId
        return this
    }

    /**
     * 设置文件名称过滤器
     *
     * > 文件过滤器、文件名称过滤器两者只能设置其一
     * > 两个都设置时，以文件过滤器为准
     */
    fun setFileNameFilter(value: FilenameFilter): EasyFileChoose {
        nameFilter = value
        return this
    }

    /**
     * 设置文件过滤器
     *
     * > 文件过滤器、文件名称过滤器两者只能设置其一
     * > 两个都设置时，以文件过滤器为准
     */
    fun setFileFilter(value: FileFilter): EasyFileChoose {
        fileFilter = value
        return this
    }

    /**
     * 自定义文件排序
     */
    fun setComparator(value: Comparator<File>): EasyFileChoose {
        comparator = value
        return this
    }

    /**
     * 开始选择文件
     */
    fun choose(activity: Activity, requestCode: Int) {
        FileChooseActivity.start(activity, requestCode)
    }

    /**
     * 获取选择的路径
     */
    fun getPath(intent: Intent): String {
        return if (intent.hasExtra(INTENT_KEY_PATH)) intent.getStringExtra(INTENT_KEY_PATH) else ""
    }

    internal fun icon(file: File): Int = if (file.isDirectory) directoryIcon else fileIcon

    internal fun backgroundColor(selected: Boolean): Int {
        return if (selected) selectedColor else transparentColor
    }
}
